src/go/configgenerator/filter_generator.go

```diff
@@ -28,68 +28,50 @@ type FilterGenerator interface {
 	// FilterName returns the name of the filter.
 	FilterName() string
 
-	// GenFilterConfig generates the filter config and a list of methods that need per-route configs.
-	GenFilterConfig(*ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error)
+	// IsEnabled returns true if the filter config should be generated.
+	// If false, none of the generation methods will be called.
+	IsEnabled() bool
+
+	// GenFilterConfig generates the filter config.
+	//
+	// Return (nil, nil) if the filter has no listener-level config, but may
+	// have per-route configurations.
+	GenFilterConfig(*ci.ServiceInfo) (*hcmpb.HttpFilter, error)
 
 	// GenPerRouteConfig generates the per-route config for the given HTTP route.
-	// This method is only called on the routes that `GenFilterConfig` returns.
+	//
+	// This method is called on all routes. Return (nil, nil) to indicate the
+	// filter does NOT require a per-route config for the given route.
 	GenPerRouteConfig(*ci.MethodInfo, *httppattern.Pattern) (*anypb.Any, error)
 }
 
 // MakeFilterGenerators provide of a slice of FilterGenerator in sequence.
 func MakeFilterGenerators(serviceInfo *ci.ServiceInfo) ([]FilterGenerator, error) {
-	var filterGenerators []FilterGenerator
-
-	if serviceInfo.Options.CorsPreset == "basic" || serviceInfo.Options.CorsPreset == "cors_with_regex" {
-		filterGenerators = append(filterGenerators, &filtergen.CORSGenerator{})
-	}
-
-	// Add Health Check filter if needed. It must behind Path Matcher filter, since Service Control
-	// filter needs to get the corresponding rule for health check cmake depend.insalls, in order to skip Report
-	if serviceInfo.Options.Healthz != "" {
-		filterGenerators = append(filterGenerators, &filtergen.HealthCheckGenerator{})
-	}
+	return []FilterGenerator{
+		filtergen.NewCORSGenerator(serviceInfo),
 
-	if serviceInfo.Options.EnableResponseCompression {
-		filterGenerators = append(filterGenerators, &filtergen.CompressorGenerator{
-			CompressorType: filtergen.GzipCompressor,
-		})
-		filterGenerators = append(filterGenerators, &filtergen.CompressorGenerator{
-			CompressorType: filtergen.BrotliCompressor,
-		})
-	}
+		// Health check filter is behind Path Matcher filter, since Service Control
+		// filter needs to get the corresponding rule for health check in order to skip Report
+		filtergen.NewHealthCheckGenerator(serviceInfo),
+		filtergen.NewCompressorGenerator(serviceInfo, filtergen.GzipCompressor),
```

Comment:
On a second thought, are we able to abstract the filter list as well? Then for the mode don't need these filters, we don't need to bother import the code even.

Reply:
Not sure I am understanding.

My thinking is that for each additional modes, we just create a new `MakeFilterGenerators` function internally. The function returns only the imported grpc_web + grpc_transcoder and new ext_auth + ext_proc filters generators.

Reply:
`MakeFilterGenerators` for each new mode SG. I thought we are going to use one `MakeFilterGenerators` to include all the filters for every mode, and isEnabled is used to determine the final result. But if we have `MakeFilterGenerators` for different mode, we don't need to bother import the filter gen code we don't need.

Reply:
Yup exactly :) The `isEnabled()` is actually to keep pre-existing behavior. I don't want `GenFilterConfig()` and `GenPerRouteConfig()` to be called if the filter is not enabled.